numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -3981,6 +3981,9 @@ array_shares_memory_impl(PyObject *args, PyObject *kwds, Py_ssize_t default_max_
     }
     else if (PyLong_Check(max_work_obj)) {
         max_work = PyLong_AsSsize_t(max_work_obj);
+        if (PyErr_Occurred()) {
```

Comment:
LGTM, two small comments (which I don't care much about)
 1.  Could use `PyNumber_Index` here, which is a bit more general then the long check, but also forbids floats, etc.
 2. Could make it `if ((max_work == -)1 && (PyErr_Occured())) {` AFAIK.